policy-service/src/main/java/uk/gov/gchq/palisade/service/policy/service/PolicyServiceHierarchyProxy.java

```diff
@@ -42,101 +41,174 @@
+public class PolicyServiceHierarchyProxy {
     private static final Logger LOGGER = LoggerFactory.getLogger(PolicyServiceHierarchyProxy.class);
 
-    final PolicyService service;
+    private final PolicyServiceCachingProxy service;
 
-    public PolicyServiceHierarchyProxy(final PolicyService service) {
+    /**
+     * Instantiates a new Policy service hierarchy proxy taking a PolicyService as an argument.
+     *
+     * @param service {@link PolicyService} used to instantiate this class
+     */
+    public PolicyServiceHierarchyProxy(final PolicyServiceCachingProxy service) {
         this.service = service;
     }
 
-    @Override
-    public <R extends Resource> Optional<R> canAccess(final User user, final Context context, final R resource) {
-        LOGGER.debug("Determining access: {} for user {} with these resources {}", context, user, resource);
-        Optional<R> serviceCanAccess = service.canAccess(user, context, resource);
-        // If the service says we can access the resource naively, check up the resource hierarchy
-        // If all parent, grandparent etc. resources say we can access the resource, then it is accessible
-        Optional<Rules<Resource>> accessRules = serviceCanAccess.flatMap(res -> getHierarchicalRules(res, Policy::getResourceRules));
-        return (Optional<R>) accessRules.map(rule -> Util.applyRulesToItem(resource, user, context, rule));
+    private static <T extends Serializable> Rules<T> mergeRules(final Rules<T> inheritedRules, final Rules<T> newRules) {
+        LOGGER.debug("inheritedRules and newRules both present\n MessageInherited: {}\n MessageNew: {}\n RulesInherited: {}\n RulesNew: {}",
+                inheritedRules.getMessage(), newRules.getMessage(), inheritedRules.getRules(), newRules.getRules());
+        Rules<T> mergedRules = new Rules<>();
+
+        // Merge messages
+        ArrayList<String> messages = new ArrayList<>();
+        String inheritedMessage = inheritedRules.getMessage();
+        String newMessage = newRules.getMessage();
+        if (!inheritedMessage.equals(Rules.NO_RULES_SET)) {
+            messages.add(inheritedMessage);
+        }
+        if (!newMessage.equals(Rules.NO_RULES_SET)) {
+            messages.add(newMessage);
+        }
+        mergedRules.message(String.join(",", messages));
+        LOGGER.debug("Merged messages: {} + {} -> {}", inheritedRules.getMessage(), newRules.getMessage(), mergedRules.getMessage());
+
+        // Merge rules
+        mergedRules.addRules(inheritedRules.getRules());
+        mergedRules.addRules(newRules.getRules());
+        LOGGER.debug("Merged rules: {} + {} -> {}", inheritedRules.getRules(), newRules.getRules(), mergedRules.getRules());
+
+        return mergedRules;
+    }
+
+    /**
+     * This method is used to find out if the given user is allowed to access
+     * the resource given their purpose. This is where any resource level
+     * access controls are enforced.
```

Comment:
I wouldnt be so specific on it being where the purpose check is applied. Since the purpose may be checked on a per record level as well. Also other contextual or user attributes my impact the decision at a resource level about if a request can proceed for this resource.